Fix locked accounts

OpenLedgerApp opened the pull request:
Prevent to create cycled authorities. Some unit tests added.

sschiessl-bcp commented:
Can you please point me to the issue # behind this PR?

abitmore commented:
* https://github.com/bitshares/bitshares-core/issues/269
* https://github.com/bitshares/bsips/issues/94
* https://github.com/bitshares/bsips/pull/149

sschiessl-bcp commented:
From reading the BSIPs and issues I am confused what is done and still needs doing. A massive amount of hours is invested here. 

Please clarify

OpenLedgerApp commented:
> From reading the BSIPs and issues I am confused what is done and still needs doing. A massive amount of hours is invested here.
> 
> Is there a roadmap or something similar for this task?

@Stefan, We had tasks list for this BSIP. 

At the first step, we have made research to identify locked accounts.
Then we have summarised the most liquid currencies of locked accounts.
At the second step, we have implemented  "Prevent to create cycled authorities".
At the third step, we offered two ways how to fix locked accounts:

1. Automatic (unconditional)
- Detect all locked accounts in first maintenance after hardfork.
- Find latest operation ("lock"-operation) among operations of cycled accounts. Do it for each cycle.
- Undo all "lock"-operations. Find previous account update/create an operation that changes authorities and redo it.

2. Manual (on-demand)
- Check if an account is locked.
- Find latest operation ("lock"-operation) in cycle.
- Undo "lock"-operation.

We have implemented to unlock accounts manually. It's a better idea because we don't know exactly if locking of accounts was by mistake or not.

In conclusion, all Pull Request with the smoke and unit tests we published on GitHub. We have finished all task for this BSIP according to our specification. 

However, we have only one open question "how to take 10% from each account as payment for the implementation".

sschiessl-bcp commented:
Can you comment how exactly the manual way should work?

OpenLedgerApp commented:
@sschiessl-bcp, we see the following use case for "manual fix of locked accounts":

Preconditions:
A User should have a valid authority which preceded account lock state. 

Steps:

1. User creates unlock_account_operation with account_to_unlock as parameter
2. User signs the operation with his previous authority 

Expected Results:
Account authority is returned to its previous state

dayman32 commented:
Can I unlock it manually now? or Do you still need hardfork?

sschiessl-bcp commented:
> Can I unlock it manually now? or Do you still need hardfork?

This is ongoing development and a core change, don't expect anything before autumn.

dayman32 commented:
Has anything changed? Can I somehow withdraw funds from blocked accounts?

pmconrad commented:
> Has anything changed? Can I somehow withdraw funds from blocked accounts?

No.

> When is Hardfork?

The next planned hardfork will not include a fix for this. The relevant BSIP is still being written. Even the BSIPs that were completed in the past month have not been voted in, so don't hold your breath.

dayman32 commented:
Another 1 year has passed. Anyone looking for a solution?
